src/controller/deletecontent.js

```diff
@@ -132,6 +132,13 @@ function mergeBranches( batch, startPos, endPos ) {
 		// To then become:
 		// <a><b>xy</b>[]</a><c>{}</c>
 		batch.merge( startPos );
+
+		// We need to check and strip disallowed attributes in direct children because after merge
+		// some attributes could end up in a parent where are disallowed.
+		//
+		// e.g. bold is disallowed for <H1>
+		// <h1>Fo{o</h1><p>b}a<b>r</b><p> -> <h1>Fo{}a<b>r</b><h1> -> <h1>Fo{}ar<h1>.
+		removeDisallowedAttributes( batch, Array.from( startParent.getChildren() ), [ startParent ] );
```

Comment:
Theoretically, `schemaPath` equal to `[ startParent ]` might be incorrect, but this is a pretty edge case and I can't think of convincing real world scenario.

Anyway, it might return wrong result if, for example `attr` is allowed on element `A` and `B` but disallowed on `B` if `B` is in `X`.

Then: `<x><b>foo</b><a><$text attr=true>bar</$text></x>` upon merging `attr` should be cleared from text and it probably would not be now.

Reply:
You can pass a `Position` instance as `inside` parameter. Maybe that would be easier and even more clear in implementation. If that would work too, it might solve the issue I described.

Reply:
One more note - I am not sure about checking only direct children. What if one of children has an element or text with incorrect attribute?

Reply:
Those also apply to the other `removeDisallowedAttributes` function.